design/automatic-network-configuration.md

```diff
@@ -0,0 +1,801 @@
+of the physical host is managed through a Kubernetes API, it would be
+convenient to be able to reconfigure related network devices using a
+similar API.
+
+## Goals
+
+- Define a Kubernetes API for configuring network device.
+- Automatically configure the network infrastructure for a host when
+adding it to a cluster.
+- Do network configuration when deprovisioning the hosts.
+- Design a network abstraction that can represent any of the target networking
+configuration, independently of the controller used.
+
+## Non-Goals
+
+- Implement a network controller which is aimed at being an integration with
+an existing controller only.
+- Implement a solution for a specific underlying infrastructure.
+- Implement a solution for link aggregation in the future.
+- Implement a solution for lldp in the future.
+
+## Proposal
+
+This document proposes to add a new mechanic to automatically perform physical
+network device configuration before provisioning a BareMetalHost. First we
+added the NetworkConfiguration operator to implement the network configuration
+of the device. Then we added some methods to CAPM3 to implement together. When
+network configuration is required, CAPM3 will process some content through
+additional methods, and then trigger the NetworkConfiguration operator to
+perform network configuration.
+
+### User Stories
+
+#### Story 1
+
+As a consumer of Metal³, when adding a machine resource to the cluster, according
+to the network configuration of the cluster, the physical host is connected to the
+corresponding logical layer-2 network segment.
+
+#### Story 2
+
+As a consumer of Metal³, when a machine is deleted from the cluster, the
+corresponding network configuration is automatically deleted.
+
+#### Story 3
+
+As a user/admin a SmartNIC that is hosted in a BMH needs to be provisioned,
+pre-configured, and then capable of in operation management.
+
+Physical server(OS) -----> SmartNIC [virtual switch(OVS) ---> physical NIC]
+←-- (LAN cable)-----> PortID
+
+#### Story 4
+
+As a user, when creating a network configuration and specifying link
+aggregation, the related BMH and switch can be automatically configured.
+
+#### Story 5
+
+As a user, when creating a network configuration, it is possible to configure ACL
+settings in a switch for each bare metal server according to its functionality.
+The ACL setting:
+
+- Direction: ingress or egress
+- Protocol: SSH, HTTP, …all
+- Action: allow, deny
+- Port status: up, down (status of the switch port)
+
+## Design Details
+
+### Structure
+
+In the network operator, we abstract the following roles:
+
+#### Device
+
+Abstract all devices in the cluster (such as Switch, SmartNic).
+Each type of network device must have its own CRD.
+
+At the same time, each device has its own specific interface, and
+needs to implement corresponding interfaces for different vendors
+and device types. The following is an example of a switch interface:
```

Comment:
We may not describe it clearly here. What we want to express is that each specific network device has its own CRD, and has its own interface, and some corresponding methods are written in the interface. After implementing this method, it can be called by the Switch instance.

Reply:
What I think you're saying is that the connection/authentication data is stored in a separate CRD but the actual implementation is not done by a separate controller for the provider CRD, instead the Switch controller calls the method on some object which must be hard-coded into the Switch controller?

This eliminates the benefit of using the Cluster API model of separate provider CRDs (e.g. Metal3Machine), which is that you can add new providers without modifying the code of the main controller (Machine in the case of Cluster API, Switch in this case). There's very little difference between what you appear to have and just including the connection/authentication data in a free-form field in the main Switch API (which is how the Cluster API originally did it in v1alpha1).

Reply:
There is a problem with what I said yesterday. 
In our proposal, the configuration of each device may have multiple back-end implementations. 
For example, in the case of Switch, we use ansible as its back-end implementation, but there may be other back-end implementations to configure the switch in the future. This requires that no matter which method is added, this method must implement the `Switch interface`, so that you do not need to modify the controller code and directly use this new method to configure the switch.
I will add a detailed description in the proposal later.

Reply:
You _do_ need to modify the controller code to support another backend, because something needs to import the new backend and call [`Register()`](https://github.com/Hellcatlk/network-operator/blob/master/pkg/backends/switches/switches.go#L21-L23) with it.

The (newer) Cluster API approach is to have a separate controller for each platform (e.g. AnsibleSwitch) so that new controllers can be added without modifying the core (Switch) controller at all. They communicate only through the CRD interfaces, not through shared golang code.

Reply:
> What I think you're saying is that the connection/authentication data is stored in a separate CRD but the actual implementation is not done by a separate controller for the provider CRD, instead the Switch controller calls the method on some object which must be hard-coded into the Switch controller?

Yes, your understanding is correct. I will revise this document to make it clearer.